src/Microsoft.AspNetCore.Http.Connections/Internal/HttpConnectionContext.cs

```diff
@@ -179,10 +180,9 @@ public void TickHeartbeat()
         {
             var disposeTask = Task.CompletedTask;
 
+            await StateLock.WaitAsync();
```

Comment:
Since we're only waiting on the statelock, does this mean that dispose can start while writes are in progress?

Reply:
We but we don't care about that race. It does mean the client will 500, which we could consider catching...

Reply:
Ok, Could it be prevented full stop by using `await Task.WhenAll(StateLock.WaitAsync(), WriteLock.WaitAsync());` and then releasing both? Or do you think it is not worth the complexity?

Reply:
"Nah" - @davidfowl 2018